plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java

```diff
@@ -0,0 +1,251 @@
+          // not containing needed elements, just skip it but this time log it
+          log.info("Malformed package.json in {}", file.getRepositoryItemUid());
+
+          continue;
+        }
+        final Version version = Version.parse(String.valueOf(versionJson.get("version")));
+        final Map<String, Object> dist = Maps.newHashMap();
+        dist.put("tarball", "generated-on-request"); // We lost original value, but does not matter as is NX hosted
+        dist.put("shasum", file.getRepositoryItemAttributes().get(StorageFileItem.DIGEST_SHA1_KEY));
+        versionJson.put("dist", dist);
+        versions.put(version, versionJson);
+      }
+      catch (Exception e) {
+        log.info("Failed to extract or malformed package.json from {}", item.getRepositoryItemUid());
+      }
+    }
+
+    // ie. dir empty or all extract failed?
+    if (versions.isEmpty()) {
+      return;
+    }
+
+    // rebuild package root as possible and push it into database
+    final Map<String, Object> rootJson = Maps.newHashMap();
+    rootJson.put("dist-tags", Maps.<String, Object>newHashMap());
+    rootJson.put("versions", Maps.<String, Object>newHashMap());
+    rootJson.put("x-nx-rebuilt", Iso8601Date.format(new Date()));
+
+    for (Version version : versions.keySet()) {
+      try {
+        processPackageJson(rootJson, versions.get(version));
+      }
+      catch (IllegalArgumentException e) {
+        log.info("Failed to process package.json from {} version {}", coll.getPath(), version, e);
+      }
+    }
+
+    // ie. all processPackageJson failed?
+    if (((Map<String, Object>) rootJson.get("versions")).isEmpty()) {
+      return;
+    }
+
+    final Map<String, Object> lastPackage = versions.get(versions.lastKey());
+    final String lastVersion = String.valueOf(lastPackage.get("version"));
+    final Map<String, Object> distTags = (Map<String, Object>) rootJson.get("dist-tags");
+    distTags.remove(lastVersion);
+    distTags.put("latest", lastVersion);
+
+    final PackageRoot packageRoot = new PackageRoot(context.getRepository().getId(), rootJson);
+    hostedMetadataService.consumePackageRoot(packageRoot);
+    packageRoots = packageRoots + 1;
+    packageVersions = packageVersions + versions.size();
+  }
+
+  private Map<String, Object> extractPackageJson(final StorageFileItem file) throws IOException {
+    try (final GZIPInputStream gzipInputStream = new GZIPInputStream(file.getInputStream())) {
+      log.debug("Examining TAR file");
+      final TarArchiveInputStream tarArchiveInputStream = new TarArchiveInputStream(gzipInputStream);
+      TarArchiveEntry tarEntry;
+      // npm uses tar option "--strip-components=1", so we need to strip first path element to have proper names
+      String tarEntryName;
+      do {
+        tarEntry = tarArchiveInputStream.getNextTarEntry();
+        tarEntryName = null;
+        if (tarEntry != null) {
+          // tar entry names are "pathelem1/pathelem2/.../filename"
```

Comment:
Some of these methods are quite long; this seems like a candidate for a clearly named method that just parses out tar entry names (as the algorithm could then be unit-tested on its own, etc.)